tapa/common/graph.py

```diff
@@ -86,3 +93,101 @@ def get_flatten_graph(self) -> "Graph":
+        new_obj["level"] = "upper"
+
+        # Find all task instances
+        insts = self.get_top_task_inst().get_subtasks_insts()
+        for inst in insts:
+            # make sure top has been flattened
+            assert isinstance(inst.definition, TaskDefinition)
+            assert inst.definition.get_level() == TaskDefinition.Level.LEAF, (
+                "Top task must be flattened for floorplanning"
+            )
+
+        # filter out tasks that are not in the slot
+        new_insts = [inst for inst in insts if inst.name in tasks_in_slot]
+
+        assert new_insts, [inst.name for inst in insts]
+
+        # obj['tasks']:
+        # construct the task insts of the slot
+        new_obj["tasks"] = defaultdict(list)
+        for inst in new_insts:
+            new_obj["tasks"][inst.definition.name].append(inst.to_dict())
+
+        # obj['fifos']:
+        # construct the fifos of the slot
+        fifos = self.get_top_task_inst().get_interconnect_insts()
+        new_fifos: list[InterconnectInstance] = []
+        fifo_ports: list[str] = []
+        for fifo in fifos:
+            fifo_obj = fifo.to_dict()
+            src = fifo_obj["consumed_by"]
+            dst = fifo_obj["produced_by"]
+            # For fifo connecting task insts inside the slot, keep it
+            if (
+                INST_NAME.format(src[0], src[1]) in tasks_in_slot
+                and INST_NAME.format(dst[0], dst[1]) in tasks_in_slot
+            ):
+                new_fifos.append(fifo)
+            # For fifo connecting a task inst inside and an inst outside the slot,
+            # put it to ports later
+            elif (
+                INST_NAME.format(src[0], src[1]) in tasks_in_slot
+                or INST_NAME.format(dst[0], dst[1]) in tasks_in_slot
+            ):
+                fifo_ports.append(fifo.name)
+        new_obj["fifos"] = {fifo.name: fifo.to_dict() for fifo in new_fifos}
+
+        # obj['ports']:
+        # Reconstruct the ports of the slot
+        # Remove all ports that are not used by the insts in the slot
+        used_args: set[str] = set()
+        for inst in new_insts:
+            assert isinstance(inst.obj["args"], dict)
+            args = inst.obj["args"]
+            assert isinstance(args, dict)
+            for arg in args.values():
+                assert isinstance(arg, dict)
+                assert isinstance(arg["arg"], str)
+                used_args.add(arg["arg"])
+        assert isinstance(new_obj["ports"], list)
+        new_ports = [port for port in new_obj["ports"] if port["name"] in used_args]
+
+        # Add fifos connecting the slot and the outside as ports
+        # Find the port on inst that connects to the fifo and copy
+        # the port to the slot
+        new_ports += self._get_used_ports(new_insts, fifo_ports)
+
+        new_obj["ports"] = new_ports
+
+        return TaskDefinition(slot_name, new_obj, top)
+
+    @staticmethod
```

Comment:
Could you just make `_get_used_ports` a module-private function (remove `@staticmethod` and the corresponding intendation)?